source/dataman/AMLightweightScanInfoCollection.cpp

```diff
@@ -43,512 +44,84 @@ int AMLightweightScanInfoCollection::indexOf(AMLightweightScanInfo *scan, int fr
-		// one at a time. However, for optimization reasons we still assume that
-		// the scan whose thumbnails are being added is the last one which was updated (lastUpdatedScanId_)
+		emit scanAboutToBeAdded(scanInfos_.count() - 2);
+		AMLightweightScanInfo* addedScanInfo = AMLightweightScanInfoFactory::single(id, database_);
 
-		AMLightweightScanInfo* matchedScanInfo = 0;
-		int lightweightScanInfoIndex = -1;
-		bool lightweightScanInfoFound = false;
-		for (int iScanInfo = 0, size = scanInfos_.count(); iScanInfo < size && !lightweightScanInfoFound; iScanInfo++)
-		{
-			if(scanInfos_.at(iScanInfo)->id() == lastUpdatedScanId_)
-			{
-				lightweightScanInfoIndex = iScanInfo;
-				matchedScanInfo = scanInfos_.at(iScanInfo);
-				lightweightScanInfoFound = true;
-			}
+		if(addedScanInfo) {
+			scanInfos_.append(addedScanInfo);
 		}
+		emit scanAdded();
+	} else if(tableName == "ObjectExperimentEntries") {
+		int updatedScanId = AMLightweightScanInfoFactory::scanIdForExperimentEntry(id, database_);
 
-		if(!matchedScanInfo)
+		if(updatedScanId == -1)
 			return;
 
-		emit scanThumbnailAboutToBeAdded();
-		matchedScanInfo->addThumbnailId(id);
-		emit scanThumbnailAdded();
-		emit scanUpdated(lightweightScanInfoIndex);
-	}
-	else if(sampleNameMap_.keys().contains(tableName))
-	{
-		populateSingleSample(tableName, id);
+		int dirtyScanInfoIndex = indexOfScanWithId(updatedScanId);
+		if(dirtyScanInfoIndex >= 0) {
+			AMLightweightScanInfo* dirtyScanInfo = scanInfos_.at(dirtyScanInfoIndex);
+			AMLightweightScanInfoFactory::updateSingle(dirtyScanInfo, database_);
+
+			emit scanUpdated(dirtyScanInfoIndex);
+		}
+	} else if(tableName == AMDbObjectSupport::s()->tableNameForClass("AMRun")) {
+		emit runAboutToBeAdded();
+		runMapCache_ = AMLightweightScanInfoFactory::runMap(database_);
+		emit runAdded();
 	}
 }
 
 void AMLightweightScanInfoCollection::onDbItemUpdated(const QString &tableName, int id)
 {
-	if(tableName == AMDbObjectSupport::s()->tableNameForClass("AMScan"))
-	{
-		// A scan has been changed. We need to find the scan that matches the id, update the corresponding
-		// scan info and emit our updated signal
-		AMLightweightScanInfo* matchedScanInfo = 0;
-		int lightweightScanInfoIndex = -1;
-		bool lightweightScanInfoFound = false;
-		for (int iScanInfo = 0, size = scanInfos_.count(); iScanInfo < size && !lightweightScanInfoFound; iScanInfo++)
-		{
-			if(scanInfos_.at(iScanInfo)->id() == id)
-			{
-				lightweightScanInfoIndex = iScanInfo;
-				matchedScanInfo = scanInfos_.at(iScanInfo);
-				lightweightScanInfoFound = true;
+	if(tableName == AMDbObjectSupport::s()->tableNameForClass("AMScan")) {
+		int dirtyScanInfoIndex = indexOfScanWithId(id);
+		if(dirtyScanInfoIndex >= 0) {
+			AMLightweightScanInfo* dirtyScanInfo = scanInfos_.at(dirtyScanInfoIndex);
+			if(dirtyScanInfo->thumbnailCount() > 0) {
+				emit scanThumbnailAboutToBeRemoved(dirtyScanInfoIndex, 0, dirtyScanInfo->thumbnailCount()-1);
```

Comment:
This captures the case when a scan is saved having had a change made which will affect the thumbnails. Because they're all removed from the database and then re-added we go ahead here and refresh the whole bunch. It's a little less efficient to prompt a redraw of the thumbnail here each time the scan is saved (which is what the signal eventually triggers) - but it is significantly cleaner and less complex than the previous work around which attempted to listen to changes to the thumbnail table and then in some way figure out which scan that related to (as we only receive a id of -1 in that case)